File: ios/Classes/InterIsolateEventChannelPlugin.swift
import Flutter
import UIKit

// Singleton class for broadcasting events across all isolates/engines
class InterIsolateBroadcaster {
  static let shared = InterIsolateBroadcaster()

  // Serial queue for thread safety
  private let queue = DispatchQueue(label: "com.joel.inter_isolate_event_channel.broadcaster")

  // Registered event sinks (managed by UUID for precise control)
  private var sinks: [UUID: FlutterEventSink] = [:]

  private init() {}

  // Registers a new event sink - returns UUID for later removal
  func addSink(_ sink: @escaping FlutterEventSink) -> UUID {
    let id = UUID()
    queue.sync {
      sinks[id] = sink
    }
    return id
  }

  // Removes a specific sink (by UUID)
  func removeSink(id: UUID) {
    queue.sync {
      sinks.removeValue(forKey: id)
    }
  }

  // Broadcasts an event to all registered sinks
  func broadcastEvent(_ event: [String: Any]) {
    queue.sync {
      for sink in sinks.values {
        // Send event on main queue
        DispatchQueue.main.async {
          sink(event)
        }
      }
    }
  }
}

public class InterIsolateEventChannelPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private var currentSinkId: UUID?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = InterIsolateEventChannelPlugin()

    // Method channel for emitting events
    let emitChannel = FlutterMethodChannel(name: "inter_isolate_event/emit", binaryMessenger: registrar.messenger())
    registrar.addMethodCallDelegate(instance, channel: emitChannel)

    // Event channel for broadcasting
    let eventChannel = FlutterEventChannel(name: "inter_isolate_event/broadcast", binaryMessenger: registrar.messenger())
    eventChannel.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    if call.method == "emitEvent", let args = call.arguments as? [String: Any] {
      guard let eventType = args["eventType"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "eventType is required", details: nil))
        return
      }

      let payload = args["payload"]
      let event: [String: Any] = [
        "eventType": eventType,
        "payload": payload as Any
      ]

      InterIsolateBroadcaster.shared.broadcastEvent(event)
      result(nil)
    } else {
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - FlutterStreamHandler methods
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    // Store returned UUID for later removal
    currentSinkId = InterIsolateBroadcaster.shared.addSink(events)
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    // Remove sink by UUID to prevent memory leaks
    if let id = currentSinkId {
      InterIsolateBroadcaster.shared.removeSink(id: id)
      currentSinkId = nil
    }
    return nil
  }
}

File: android/src/main/kotlin/com/joel/inter_isolate_event_channel/inter_isolate_event_channel/InterIsolateEventChannelPlugin.kt
package com.joel.inter_isolate_event_channel.inter_isolate_event_channel

import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.CopyOnWriteArraySet

/** InterIsolateEventChannelPlugin */
class InterIsolateEventChannelPlugin: FlutterPlugin, EventChannel.StreamHandler {
  // Event channel for broadcasting
  private lateinit var eventChannel: EventChannel
  // Method channel for emitting events
  private lateinit var emitChannel: MethodChannel
  // Current registered sink (for removal on cancel)
  private var currentSink: EventChannel.EventSink? = null

  // Singleton instance management
  companion object {
    private val instance = InterIsolateBroadcaster()

    fun getInstance(): InterIsolateBroadcaster {
      return instance
    }
  }

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    // Method channel for emitting events
    emitChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "inter_isolate_event/emit")
    emitChannel.setMethodCallHandler { call, result ->
      if (call.method == "emitEvent") {
        val eventType = call.argument<String>("eventType")
        val payload = call.argument<Any>("payload")

        if (eventType != null) {
          val event = mapOf(
            "eventType" to eventType,
            "payload" to payload
          )
          getInstance().broadcastEvent(event)
          result.success(null)
        } else {
          result.error("INVALID_ARGUMENT", "eventType is required", null)
        }
      } else {
        result.notImplemented()
      }
    }

    // Event channel for broadcasting
    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "inter_isolate_event/broadcast")
    eventChannel.setStreamHandler(this)
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    emitChannel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
  }

  // EventChannel.StreamHandler implementation
  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    events?.let {
      currentSink = it
      getInstance().addSink(it)
    }
  }

  override fun onCancel(arguments: Any?) {
    // Explicitly remove sink to prevent memory leaks
    currentSink?.let {
      getInstance().removeSink(it)
      currentSink = null
    }
  }
}

/**
 * Singleton class for broadcasting events across all isolates/engines
 */
class InterIsolateBroadcaster {
  // Thread-safe event sink collection
  private val sinks = CopyOnWriteArraySet<EventChannel.EventSink>()

  /**
   * Registers a new event sink
   */
  fun addSink(sink: EventChannel.EventSink) {
    sinks.add(sink)
  }

  /**
   * Removes an event sink
   */
  fun removeSink(sink: EventChannel.EventSink) {
    sinks.remove(sink)
  }

  /**
   * Broadcasts an event to all registered sinks
   */
  fun broadcastEvent(event: Map<String, Any?>) {
    // Send event to all sinks
    val iterator = sinks.iterator()
    while (iterator.hasNext()) {
      val sink = iterator.next()
      try {
        sink.success(event)
      } catch (e: Exception) {
        // Remove failed sinks immediately (CopyOnWriteArraySet allows removal during iteration)
        sinks.remove(sink)
      }
    }
  }
}
